### DISCUSSION TITLE
emitting an effect upon completion of a publisher

### DISCUSSION THREAD

**Discussion by falcowinkler** [pointfreeco/swift-composable-architecture#990] [Q&A] [answered]
I have an `Effect` in my environment that does some long term data processing.
It publishes many results over time.
Since i sometimes need to react on intermediate results of this data processing, i feed the results back into my reducer via an action.

Now i also want to react on the event that this processing is complete. In combine, i would simply handle the completion event
(via `handleEvents`). 
But since i have to open up a closure here, i of course can not return an effect from the reducer in this case.

I could  of course refactor the implementation of my data processing to emit an enum of the value that i care about, and a special `completion` case, that i then wire up via a `sink` (transparent to the reducer). But it's not a very elegant solution..

Did you experience such a requirement and if so, how did you solve it?

**Comment by stephencelis** [comment, marked as answer]
@falcowinkler If you concatenate an effect to your long-running effect, it will emit when the long-running effect completes:

```swift
return Effect.concatenate(
  myLongRunningEffect,
  Effect(value: MyAction.myLongRunningEffectCompleted)
)
```
